runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/util/BoundedQueueExecutor.java

```diff
@@ -236,10 +241,37 @@ private void executeMonitorHeld(Runnable work, long workBytes) {
   }
 
   private void decrementCounters(long workBytes) {
-    monitor.enter();
-    --elementsOutstanding;
-    bytesOutstanding -= workBytes;
-    monitor.leave();
+    // All threads queue decrements and one thread grabs the monitor and updates
+    // counters. We do this to reduce contention on monitor which is locked by
+    // GetWork thread
+    decrementQueue.add(workBytes);
+    boolean submittedToExistingBatch = decrementBatchOpen.getAndSet(true);
+    if (submittedToExistingBatch) {
+      // There is already a thread about to drain the decrement queue
+      // Current thread does not need to drain.
+      return;
+    }
+    synchronized (decrementQueueDrainLock) {
+      decrementBatchOpen.set(false);
```

Comment:
Since it's a little subtle how about a comment.

// By setting false here, we may allow another decrement to claim submission of the next batch and start waiting on the decrementQueueDrainLock.
// However this prevents races that would leave decrements in the queue and unclaimed and we are ensured there is at most one additional thread blocked. This helps prevent the executor from creating threads over the limit if many were contending on the lock while their decrements were already applied.